File: 2-cloudrun-quarkus-native/src/main/java/com/benchmark/device/service/DeviceService.java
package com.benchmark.device.service;

import com.benchmark.device.model.Device;
import com.benchmark.device.repository.DeviceRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@ApplicationScoped
public class DeviceService {

    @Inject
    DeviceRepository repository;

    public Device create(Device device) {
        device.setId(UUID.randomUUID());
        device.setCreatedAt(Instant.now());
        return repository.save(device);
    }

    public Optional<Device> getById(UUID id) {
        return repository.findById(id);
    }

    public List<Device> listByStatus(String status) {
        if (status == null || status.isBlank()) {
            return repository.findAll();
        }
        return repository.findByStatus(status);
    }
}

File: 3-cloudrun-golang/service/device_service.go
package service

import (
	"time"

	"github.com/benchmark/cloudrun-go/model"
	"github.com/benchmark/cloudrun-go/repository"
	"github.com/google/uuid"
)

type DeviceService struct {
	repo *repository.DeviceRepository
}

func NewDeviceService(repo *repository.DeviceRepository) *DeviceService {
	return &DeviceService{repo: repo}
}

func (s *DeviceService) Create(req model.CreateDeviceRequest) model.Device {
	device := model.Device{
		ID:        uuid.New().String(),
		Name:      req.Name,
		Status:    req.Status,
		CreatedAt: time.Now().UTC(),
	}
	return s.repo.Save(device)
}

func (s *DeviceService) GetByID(id string) (model.Device, bool) {
	return s.repo.FindByID(id)
}

func (s *DeviceService) ListByStatus(status string) []model.Device {
	if status == "" {
		return s.repo.FindAll()
	}
	return s.repo.FindByStatus(status)
}
